File: lib/state-machine.ts
import type { BotState } from "@/types";

export function normalizeProjectName(value: string) {
  return value.toLowerCase().replace(/[^a-z0-9]+/g, " ").trim().replace(/\s+/g, " ");
}

function levenshtein(a: string, b: string) {
  const matrix = Array.from({ length: a.length + 1 }, (_, i) => [i]);
  for (let j = 1; j <= b.length; j += 1) matrix[0][j] = j;

  for (let i = 1; i <= a.length; i += 1) {
    for (let j = 1; j <= b.length; j += 1) {
      matrix[i][j] =
        a[i - 1] === b[j - 1]
          ? matrix[i - 1][j - 1]
          : Math.min(matrix[i - 1][j - 1] + 1, matrix[i][j - 1] + 1, matrix[i - 1][j] + 1);
    }
  }

  return matrix[a.length][b.length];
}

export function isProjectNameMatch(candidate: string, projectName: string) {
  const a = normalizeProjectName(candidate);
  const b = normalizeProjectName(projectName);
  if (!a || !b) return false;
  if (a === b || a.includes(b) || b.includes(a)) return true;
  const distance = levenshtein(a, b);
  return distance / Math.max(a.length, b.length) <= 0.2;
}

export function parsePositiveInteger(value: string) {
  const match = value.match(/\d+/);
  if (!match) return null;
  const number = Number(match[0]);
  return Number.isInteger(number) && number > 0 && number < 200 ? number : null;
}

export function parseFloorNames(value: string, expectedCount: number) {
  const names = value
    .split(/\r?\n|,/)
    .map((line) => line.replace(/^\d+[\).\-\s]+/, "").trim())
    .filter(Boolean);

  if (names.length !== expectedCount) {
    return { ok: false as const, names, error: `Expected ${expectedCount} floor names, received ${names.length}.` };
  }

  return { ok: true as const, names };
}

export function parseBindCommand(text: string) {
  const match = text.trim().match(/^\/bind(?:@\w+)?\s+([A-Za-z0-9_-]+)\s*$/i);
  return match?.[1]?.toUpperCase() ?? null;
}

export function stateLabel(state: BotState) {
  return state
    .split("_")
    .map((part) => part[0] + part.slice(1).toLowerCase())
    .join(" ");
}

File: backend/app/state_machine.py
import re


def normalize_project_name(value: str) -> str:
    value = value.lower()
    value = re.sub(r"[^a-z0-9]+", " ", value).strip()
    return re.sub(r"\s+", " ", value)


def is_project_name_match(candidate: str, project_name: str) -> bool:
    a = normalize_project_name(candidate)
    b = normalize_project_name(project_name)
    return bool(a and b and a == b)


def is_person_name_match(candidate: str, expected: str) -> bool:
    return is_project_name_match(candidate, expected)


def parse_verification_details(value: str) -> dict:
    lines = [line.strip() for line in re.split(r"\r?\n", value) if line.strip()]
    full_name_line = next((l for l in lines if re.match(r"^full\s*name\s*:", l, re.I)), None)
    project_line = next((l for l in lines if re.match(r"^project\s*:", l, re.I)), None)
    full_name = re.sub(r"^full\s*name\s*:\s*", "", full_name_line, flags=re.I).strip() if full_name_line else (lines[0] if lines else "")
    project_name = re.sub(r"^project\s*:\s*", "", project_line, flags=re.I).strip() if project_line else (lines[1] if len(lines) > 1 else "")
    return {"fullName": full_name, "projectName": project_name}


def parse_positive_integer(value: str) -> int | None:
    match = re.search(r"\d+", value)
    if not match:
        return None
    number = int(match.group(0))
    return number if 0 < number < 200 else None


def parse_floor_names(value: str, expected: int) -> dict:
    parts = re.split(r"\r?\n|,", value)
    names = [re.sub(r"^\d+[\).\-\s]+", "", p).strip() for p in parts]
    names = [n for n in names if n]
    if len(names) != expected:
        return {"ok": False, "names": names, "error": f"Expected {expected} floor names, received {len(names)}."}
    return {"ok": True, "names": names}


def parse_bind_command(text: str) -> str | None:
    match = re.match(r"^/bind(?:@\w+)?\s+([A-Za-z0-9_-]+)\s*$", text.strip(), re.I)
    return match.group(1).upper() if match else None


def parse_start_payload(text: str) -> str | None:
    match = re.match(r"^/start(?:@\w+)?(?:\s+([A-Za-z0-9_-]+))?\s*$", text.strip(), re.I)
    return match.group(1) if match and match.group(1) else None


def normalize_telegram_username(username: str | None) -> str:
    if not username:
        return ""
    return username.strip().lstrip("@").lower()


def normalize_project_code(value: str) -> str:
    return value.strip().upper()
